user/TASK/gimbal_task/gimbal_task.c

```diff
@@ -121,24 +172,63 @@ static void get_new_data(Gimbal_t *gimbal_data){
     
-    gimbal_set->yaw_motor.pos_set += (-1) * int16_deadzone(gimbal_set->rc_update->rc.ch[2], -DEADBAND, DEADBAND) / 10;
-    gimbal_set->pitch_motor.pos_set += int16_deadzone(gimbal_set->rc_update->rc.ch[3], -DEADBAND, DEADBAND) / 10;
+    //yaw:
+    //rc -> theta -> complex rotation -> new setpoint
+
+    // TODO: FIX
+    if(gimbal_set->rc_update->rc.s[0] == RC_SW_MID || gimbal_set->rc_update->rc.s[0] == RC_SW_UP){
+        fp32 theta = -1 * int16_deadzone(gimbal_set->rc_update->rc.ch[2], -DEADBAND, DEADBAND)
+                * Motor_Ecd_to_Rad / 80.0f;
+        fp32 rotation[2] = {cos(theta), sin(theta)}; // TODO: consider alternative {cos{theta}, sin{theta}}
+        multiply_complex_a_by_b(gimbal_set->yaw_setpoint, rotation);
+        make_unit_length(gimbal_set->yaw_setpoint);
+
+        gimbal_set->pitch_motor.pos_set += int16_deadzone(gimbal_set->rc_update->rc.ch[3], -DEADBAND, DEADBAND) / 80.0f;
+    }
     
-    int16_constrain(gimbal_set->yaw_motor.pos_set, YAW_MIN, YAW_MAX);
-    int16_constrain(gimbal_set->pitch_motor.pos_set, PITCH_MIN, PITCH_MAX);
+    char print[30];
+    sprintf(print, "Constrain pitch %d between %d, %d \n\r", gimbal_set->pitch_motor.pos_set, PITCH_MIN, PITCH_MAX);
+    serial_send_string(print);
+    //int16_constrain(gimbal_set->yaw_motor.pos_set, YAW_MIN, YAW_MAX);
+    gimbal_set->pitch_motor.pos_set = int16_constrain(gimbal_set->pitch_motor.pos_set, PITCH_MIN, PITCH_MAX);
+    sprintf(print, "Constrained to %d \n\r", gimbal_set->pitch_motor.pos_set);
+    serial_send_string(print);
+}
 
-    //TODO: worry about the case where pos_set is unsigned and rc channel manages to push it    
-    // negative for a moment, causing the position to wrap from below 0 to a maxvalue. 
+static fp32 a_dot_b(fp32 a[2], fp32 b[2]){
+    return a[0] * b[0] + a[1] * b[1];
 }
 
+static fp32 length_of_a_cross_b(fp32 a[2], fp32 b[2]){
+    return a[0] * b[1] - a[1] * b[0];
+}
+
+static int16_t get_error_sign(fp32 actual_position[2], fp32 set_position[2]){
+    return length_of_a_cross_b(actual_position, set_position) >= 0 ? 1 : -1;
+}    
+
+
 /** 
  * @brief  Increments PID loop based on latest setpoints and latest positions
  * @param  None
  * @retval None
  */
 static void increment_PID(Gimbal_t *gimbal_pid){
+    /* Get error sign
+        get error val
+        insert into pid*/
+    // error magnitude 1 * ERROR_MULTIPLIER corresponds to gimbal being 90 degrees out of place --> jump action observed by pedram
+    // TODO: consider revising PID to eliminate the ERROR_MULTIPLIER constant
+    // TODO: retune PID currently very slow
+    fp32 error = get_error_sign(gimbal_pid->yaw_position, gimbal_pid->yaw_setpoint) 
+                    * (1 - a_dot_b(gimbal_pid->yaw_position,gimbal_pid->yaw_setpoint))
+                    * ERROR_MULTIPLIER;
+    // TODO: Modify pid to reject massive d term spike this could create. 
+    // TODO: Consider how this isn't quite a linear error (dot product)
```

Comment:
This last TODO is pretty easy to do, so I think should be done - others idk how easy

Reply:
Some of these require more thought or else the robot. Deleted the ones that were done